[SIL Diagnostics] Warn when assigning an integer literal to a variable of floating-point type results in loss of precision.

ravikandhadai opened the pull request:
This feature was requested in the radar issue: <rdar://15224041>.

ravikandhadai commented:
The warnings are disabled when the conversion happens through explicit constructor call. E.g. 

`$ let f3: Float = 16777217` 
`warning: '16777217' loses precision during conversion to 'Float'`
`$ Float(16777217)` // no warning

ravikandhadai commented:
@swift-ci please smoke test

stephentyrone commented:
Would it be more clear if we tell the user what value actually results? Maybe something along the lines of: "Warning: '16777217' is not exactly representable as 'Float'; 'f3' has value '16777216'.

ravikandhadai commented:
@swift-ci please smoke test

ravikandhadai commented:
@stephentyrone  Sounds good to me. I have pushed  a commit that prints the new value in the warning messages. E.g.

$ let f3: Float = 16777217
warning: '16777217' is not exactly representable as 'Float'; it becomes '16777216'

stephentyrone commented:
Nice! If you update the PR with conflicts resolved, I can review later today or tomorrow.

jrose-apple commented on test/SILOptimizer/diagnostic_constant_propagation.swift:
```diff
@@ -185,49 +185,53 @@ func testConvertOverflow() {
-  // 2^128 * (2^25-1)/2^25
+  var _ /*float32_max*/                     : Float32 = (340282346638528859811704183484516925440) // in hexfloat = 0x1.fffffep127
+
+  var _ /*float32_near_max_but_imprecise*/  : Float32 = (340282326356119256160033759537265639425) // expected-warning {{'340282326356119256160033759537265639425' is not exactly representable as 'Float32' (aka 'Float'); it becomes '3.40282326E+38'}}
```
We probably want to be printing these with full precision, to make it clear what the difference is.

ravikandhadai replied:
@jrose-apple Yeah sure, let me make that change.

ravikandhadai committed:
[SIL Diagnostics] Warn when assigning an integer literal to a
variable of floating-point type results in loss of precision.
<rdar://15224041>

ravikandhadai commented:
@jrose-apple Done. Now the new float value is displayed like an integer to make the imprecision obvious E.g. 
`var _ : Float32 = (340282326356119256160033759537265639425)`
`> warning: '340282326356119256160033759537265639425' is not exactly representable as 'Float32' (aka 'Float'); it becomes '340282326356119256160033759537265639424'`

@stephentyrone  I have fixed the merged conflicts. Thanks for your help with this.

ravikandhadai commented:
@swift-ci please smoke test

stephentyrone reviewed (approved):
LGTM, thanks!

ravikandhadai commented:
@swift-ci please test and merge

ravikandhadai commented:
@swift-ci please test and merge

shahmishal commented:
@swift-ci please test and merge

ravikandhadai commented:
@swift-ci please test and merge

ravikandhadai commented:
@swift-ci please test and merge
